### ISSUE TITLE
tmt_id and tooling around

### ISSUE THREAD

**Issue by sopos** [teemtee/tmt#1166]
We are gonna introduce the _id_ (_uuid_) of the _fmf nodes_. Such _id_ should be path-agnostic within the git repo but furthermore it could be also repo-agnostic with help of a central registry.

For that purpose I propose to call it the _tmt_id_. The purpose of such _tmt_id_ is basically to be able to translate it into the _fmf_id_ which points to an existing _fmf node_.

Following would be the possible forms of the references of:
* an _fmf_id_ inside a git repo
```yaml
url: https://..
id: f20f2608-d048-11ec-87f3-c72d591e53dd
```
* an _fmf_id_ stored in a specific key-value store
```yaml
db: http://...
id: f20f2608-d048-11ec-87f3-c72d591e53dd
```
* an _fmf_id_ stored in a central registry configured in the _tmt_ configuration
```yaml
id: f20f2608-d048-11ec-87f3-c72d591e53dd
```
The last two are basically the same except where the db information comes from.

Such _id-only_ references could be then easily used in the ReportPortal, Polarion, Jira, Bugzilla, ...

The outcome of this issue should be:
1. to create an _tmt_id_ spec in _tmt_
2. to teach _tmt_ to use such _id_ as an `require` or `recommend`
3. to create another project under https://github.com/teemtee/ which would cover the DB (maybe just an ansible playbook)
    this could be extended in the future to hold complete metadata of the _fmf nodes_ to act as an web presentation layer 
4. to teach _tmt_ to export the data to the key-value store (`tmt tests export --how DB`)

Let's discuss it and polish the scope.

**Comment by psss** [comment]
Thanks for outlining the solution! Sounds good and proposes a nice answer to the question shared on the hacking session:

> How to represent test case ID in a form of a URL (for future Jira linking)?

If there would be a central database available at a known location the url could look simply like this:

* https://tmt.db.org/f20f2608-d048-11ec-87f3-c72d591e53dd

A simple caching server could provide a nice web page with details about the `tmt` object (test, plan or story) including links to the git repository web interface to allow users to quickly check the source code. Plus a machine readable format could be provided as well, for example:

* https://tmt.db.org/f20f2608-d048-11ec-87f3-c72d591e53dd.yaml

@kkaarreell, what do you think?

**Comment by sopos** [comment]
I just updated the scope with priorities proposal

**Comment by kkaarreell** [comment]
I am not sure about the central DB reference. For test case URL I would prefer something not dependent on the presence of the actual DB. E.g. something like `http://path/to/repo?ref=REF&name=/name/of/test`, even though it would be artificial.

Regarding the caching DB itself, I think it should not only rely on a case ID but it should be able to work also with `repo, ref, testname` in some form since not all tests in a monitored repo have to have ID assigned. Ideally, it would find testcase using both the ID and using the other reference (e.g. in a form of URL mentioned above).

**Comment by jharuda** [comment]
How about separating this feature on centralized database and one-liner for specifying fmf-id?

one-liner:  `http://path/to/repo?ref=REF&path=/path/of/&name=/name/of/test`

**Comment by kkaarreell** [comment]
I have just filed related RFE https://github.com/teemtee/tmt/issues/2445

**Comment by psss** [comment]
@martinhoyer agreed to update the current progress here and check what is remaining to be done.
